src/xtgeo/surface/regular_surface.py

```diff
@@ -2781,39 +2787,25 @@ def _ensure_correct_values(
             Nothing, self._values will be updated inplace
 
         """
-        if not self._isloaded:
-            return
-
-        if values is None or values is False:
-            self._values = None
-            return
-
         currentmask = None
-        if isinstance(self._values, ma.MaskedArray):
-            currentmask = ma.getmaskarray(self._values)
+        if isinstance(self.values, ma.MaskedArray):
+            currentmask = ma.getmaskarray(self.values)
 
         if isinstance(values, ma.MaskedArray):
             newmask = ma.getmaskarray(values)
-
+            vals = values.astype(np.float64)
+            vals = ma.masked_greater(vals, self.undef_limit)
+            vals = ma.masked_invalid(vals)
             if (
                 currentmask is not None
                 and np.array_equal(currentmask, newmask)
-                and self._values.shape == values.shape
+                and self.values.shape == values.shape
                 and values.flags.c_contiguous is True
```

Comment:
Is this special casing needed? I could remove four lines here and the only thing that would change is that the memory address changes (and we dont try a reshape, but that can be skipped)